File: 1834-minimum-number-of-people-to-teach/minimum-number-of-people-to-teach.cpp
class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        unordered_set<int> users;
        vector<unordered_set<int>> knows(500);
        for (auto& relation: friendships) {
            int ui = relation[0], vi = relation[1];
            if (knows[ui - 1].size() == 0) {
                for (int& lang: languages[ui - 1]) {
                    knows[ui - 1].insert(lang);
                }
            }
            bool canTalk = false;
            for (int& lang: languages[vi - 1]) {
                if (knows[ui - 1].count(lang)) {
                    canTalk = true;
                    break;                    
                }
            }

            if (!canTalk) {
                users.insert(ui - 1);
                users.insert(vi - 1);
            }
        }

        vector<int> langs(n, 0);
        for (const int& user: users) {
            for (int& lang: languages[user]) {
                langs[lang - 1]++;
            }
        }

        return users.size() - *max_element(langs.begin(), langs.end());
    }
};

File: 1834-minimum-number-of-people-to-teach/minimum-number-of-people-to-teach.java
class Solution {
    public int minimumTeachings(int n, int[][] languages, int[][] friendships) {
        Set<Integer> users = new HashSet<>();
        List<Set<Integer>> knows = new ArrayList<>();
        for (int i = 0; i < languages.length; i++) {
            knows.add(new HashSet<>());
        }
        for (int[] friends: friendships) {
            int ui = friends[0], vi = friends[1];
            if (knows.get(ui - 1).size() == 0) {
                for (int lang: languages[ui - 1]) {
                    knows.get(ui - 1).add(lang);
                }
            }
            boolean canTalk = false;
            for (int lang: languages[vi - 1]) {
                if (knows.get(ui - 1).contains(lang)) {
                    canTalk = true;
                    break;
                }
            }

            if (!canTalk) {
                users.add(ui - 1);
                users.add(vi - 1);
            }
        }

        int[] langs = new int[n];
        int maxEle = 0;
        for (Integer user: users) {
            for (int lang: languages[user]) {
                langs[lang - 1]++;
                maxEle = Math.max(maxEle, langs[lang - 1]);
            }
        }

        return users.size() - maxEle;        
    }
}